cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -0,0 +1,335 @@
+ */
+class ConstraintCheckerService extends AbstractExecutionThreadService {
+  private static final Logger LOG = LoggerFactory.getLogger(ConstraintCheckerService.class);
+
+  private final List<ConstraintCheckerThread> constraintCheckerThreads;
+  private final Transactional transactional;
+  private final DatasetFramework datasetFramework;
+  private final Store store;
+  private final ProgramLifecycleService lifecycleService;
+  private final PropertiesResolver propertiesResolver;
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+  private final CConfiguration cConf;
+  private ScheduleTaskRunner taskRunner;
+  private ListeningExecutorService taskExecutorService;
+
+  @Inject
+  ConstraintCheckerService(Store store,
+                           ProgramLifecycleService lifecycleService, PropertiesResolver propertiesResolver,
+                           NamespaceQueryAdmin namespaceQueryAdmin,
+                           CConfiguration cConf,
+                           DatasetFramework datasetFramework,
+                           TransactionSystemClient txClient) {
+    this.store = store;
+    this.lifecycleService = lifecycleService;
+    this.propertiesResolver = propertiesResolver;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.cConf = cConf;
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(datasetFramework), txClient,
+        NamespaceId.SYSTEM, ImmutableMap.<String, String>of(), null, null)),
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+    this.datasetFramework = datasetFramework;
+    this.constraintCheckerThreads = new ArrayList<>();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    taskExecutorService = MoreExecutors.listeningDecorator(
+      Executors.newCachedThreadPool(Threads.createDaemonThreadFactory("constraint-checker-task")));
+    taskRunner = new ScheduleTaskRunner(store, lifecycleService, propertiesResolver,
+                                        taskExecutorService, namespaceQueryAdmin, cConf);
+  }
+
+  @Override
+  protected void run() {
+    LOG.info("Start running ConstraintCheckerService");
+    if (!isRunning()) {
+      return;
+    }
+
+    for (int partition = 0; partition < getNumPartitions(); partition++) {
+      constraintCheckerThreads.add(new ConstraintCheckerThread(partition));
+    }
+
+    for (ConstraintCheckerThread thread : constraintCheckerThreads) {
+      thread.start();
+    }
+
+    for (ConstraintCheckerThread thread : constraintCheckerThreads) {
+      Uninterruptibles.joinUninterruptibly(thread);
+    }
+  }
+
+  private int getNumPartitions() {
+    try {
+      return Transactions.execute(transactional, new TxCallable<Integer>() {
+        @Override
+        public Integer call(DatasetContext context) throws Exception {
+          return getJobQueue(context).getNumPartitions();
```

Comment:
It doesn't interact w/ the Table, but I needed to instantiate a dataset and the simplest way was to use the DatasetContext, which was available in the Transactions method.
I suppose I could use DatasetFramework to instantiate the dataset or just make the `getNumPartitions` a static method.